src/peers/ui.py

```diff
@@ -0,0 +1,95 @@
+import base64
+from dataclasses import asdict
+import json
+from typing import Callable, Dict, List
+import websockets
+from websockets.server import WebSocketServerProtocol
+
+from opus import Opus
+
+UIEventListener = Callable[[], None]
+
+
+class UI:
+    """
+    Handles communication with UIs.
+
+    This has a websocket server that the UI can connect to.
+
+    Parameters
+    ----------
+    opus
+        The opus. Some contents are sent in the initial handshake with a client.
+    initial_history
+        The initial history
+    """
+
+    def __init__(self, opus: Opus, initial_history: List[str]):
+        self._opus = opus
+        self._history = initial_history
+        self._listeners: Dict[str, List[UIEventListener]] = {}
+        self._websockets: List[WebSocketServerProtocol] = []
+
+    def add_event_listener(self, event_name: str, listener: UIEventListener):
+        """
+        Add a listener to an event.
+
+        Parameters
+        ----------
+        event_name
+            The event to listen to
+        listener
+            A function to be executed when the event fires
+        """
+        if event_name not in self._listeners:
+            self._listeners[event_name] = []
+        self._listeners[event_name].append(listener)
+
+    def _emit(self, event_name: str):
+        """
+        Emit an event.
+
+        Parameters
+        ----------
+        event_name
+            The event to emit
+        """
+        for listener in self._listeners[event_name]:
```

Comment:
We should use get() instead of brackets here. If the event_name isn't listened on the self._listeners[event_name] will throw a KeyException :(

Reply:
Done